lib/suma/payment/account.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require "suma/moneyutil"
+require "suma/payment"
+
+class Suma::Payment::Account < Suma::Postgres::Model(:payment_accounts)
+  plugin :timestamps
+
+  many_to_one :customer, class: "Suma::Customer"
+  many_to_one :vendor, class: "Suma::Vendor"
+  one_to_many :ledgers, class: "Suma::Payment::Ledger"
+
+  def self.lookup_platform_account
+    return Suma.cached_get("platform_payment_account") do
+      pa = self[is_platform_account: true]
+      pa ||= self.create(is_platform_account: true)
+      pa
+    end
+  end
+
+  def self.lookup_platform_vendor_service_category_ledger(cat)
+    return Suma.cached_get("platform_payment_ledger_for_category_#{cat.id}") do
+      pa = self.lookup_platform_account
+      pa.lock!
+      unless (led = pa.ledgers_dataset[vendor_service_categories: cat])
+        led = pa.add_ledger({currency: Suma.default_currency})
+        led.add_vendor_service_category(cat)
+      end
+      led
+    end
+  end
+
+  def total_balance
+    return self.ledgers.sum(Money.new(0), &:balance)
+  end
+
+  def find_chargeable_ledgers(vendor_service, amount, allow_negative_balance: false)
+    raise ArgumentError, "amount must be positive, got #{amount.format}" unless amount.positive?
+    raise Suma::InvalidPrecondition, "#{self.inspect} has no ledgers" if self.ledgers.empty?
+    contributions = []
+    self.ledgers.each do |led|
+      cat = led.category_used_to_purchase(vendor_service)
+      contributions << ChargeContribution.new(ledger: led, amount: 0, category: cat) if cat
+    end
+    contributions.sort_by! { |c| [-c.category.hierarchy_depth, c.ledger.id] }
```

Comment:
Have never seen this Dash "-" syntax before. What does it do? @rgalanakis

Reply:
http://www.rubyinside.com/rubys-unary-operators-and-how-to-redefine-their-functionality-5610.html

It's the unary `-` operator. Ie, you can think of `-1` as "the Integer type declares the unary operator `-` to return `self * -1` (okay I get this is circular but work with me). So `-1` and `x = 1; -x` are equivalent.